Exception-Handling.md

```diff
@@ -0,0 +1,209 @@
+
+Here is an example of an error being handled:
+
+```ruby
+# sample_exception.rb
+begin
+  # Dividing by zero causes an error
+  quotient = 5 / 0
+  puts "Made it past the error"
+rescue
+  quotient = nil
+  puts "Rescued the error and set quotient to nil"
+end
+```
+
+The output would be:
+```bash
+$  ruby sample_exception.rb
+Rescued the error and set quotient to nil
+```
+
+Without the rescue:
+
+```ruby
+# sample_exception.rb
+begin
+  # Dividing by zero causes an error
+  quotient = 5 / 0
+  puts "Made it past the error"
+end
+```
+The output would be:
+```bash
+$  ruby sample_exception.rb
+basic_exception_without_rescue.rb:4:in `/': divided by 0 (ZeroDivisionError)
+	from basic_exception_without_rescue.rb:4:in `<main>'
+```
+
+### Rescuing Specific Types of Exceptions
+
+There are a number of kinds of exceptions both for [regular'ol Ruby](http://blog.nicksieger.com/articles/2006/09/06/rubys-exception-hierarchy/) & [Rails](https://ruby-doc.org/core-2.3.1/Exception.html).  Some types of errors we probably don't want to try to rescue because they are irrecoverable or so unexpected that we shouldn't write code to recover from them.  We can however set up specific blocks to recover from specific types of errors that we can and want to recover from.  Generally recoverable errors inherit from `StandardError` and for this reason when you place a rescue block without specifying the exception class Ruby will only handle exceptions inheriting from StandardError.  
+
+The syntax will look like this:
+
+```ruby
+begin
+  # some code that will create an exception.
+  
+rescue <some kind of exception>
+  # Code to recover from that exception
+
+rescue <some other kind of exception>
+  # Code to recover from another type of exception
+
+rescue
+  # other exceptions handled here
+
+ensure
+  #  This code runs no matter what.  
+
+end 
+```
+#### One Ruby Example - CSVReader
+
+In Ruby we could write code to handle problems, like in our CSV code to deal with problems, like being unable to open the CSV file.
+
+```ruby
+content = []
+begin
+  CSV.foreach("file.csv") do |row|
+
+    content << row
+  end
+
+  rescue NameError
+    puts "I couldn't open the file."
+
+end
+```
+
+#### Exceptions Continue Up The Stack
+
+When an exception occurs, if it is not rescued in the immediate block or method it proceeds up and out of the method until the application crashes with an error message.
+
+```ruby
+def quotient_and_remainer dividend, divisor
+  answer = [dividend / divisor]
+  answer << quotient % divisor
+  answer
+end
+
+begin
+  puts quotient_and_remainer 10, 0
+rescue ZeroDivisionError
+  puts "Attempted to divide by zero"
+end
+```
+The method `divide` here will generate an exception when it tries to divide 10 by zero.  The exception is not "caught" in the method so ruby leaves the method and falls back to the caller of the method, which does have a rescue block to catch the `ZeroDivisionError`.  
```

Comment:
I don't know if it would be more confusing or not, but perhaps we could give an example of nested rescue blocks:
```ruby
def read_and_divide(dividend, divisor_file)
  begin
    divisor = read_divisor(divisor_file)
    quotient = divide_safely(dividend, divisor)
  rescue SystemCallError
    puts "Couldn't read divisor value from #{divisor_file}"
  end

  puts "Execution continues here after success OR rescuing from SystemCallError"
  return quotient
end

def read_divisor(divisor_file)
  f = File.open(divisor_file)
  divisor = f.gets.chomp
  return Float(divisor)
end

def divide_safely(dividend, divisor)
  begin
    return dividend / divisor
  rescue ZeroDivisionError
    puts "Attempted to divide #{dividend} by zero"
    return nil
  end
end
```

And then show the different control flow paths we can get:
```ruby
# When divisor.txt has the text "1.0" in it
read_and_divide(3.0, "divisor.txt")
# Execution continues here after success OR rescuing from SystemCallError
# => 3.0

# When divisor.txt has the text "0.0" in it
read_and_divide(3.0, "divisor.txt")
# Attempted to divide 3.0 by zero
# Execution continues here after success OR rescuing from SystemCallError
# => nil

# When divisor.txt does not exist
read_and_divide(3.0, "divisor.txt")
# Couldn't read divisor value from divisor.txt
# Execution continues here after success OR rescuing from SystemCallError
# => nil

# When divisor.txt has the text "words words words" in it
read_and_divide(3.0, "divisor.txt")
# ArgumentError: invalid value for Float(): "words words words"
#	from (irb):99:in `Float'
#	from (irb):99:in `read_divisor'
#	from (irb):89:in `read_and_divide'
#	from (irb):112
#	from /Users/charles/.rubies/ruby-2.3.1/bin/irb:11:in `<main>'
```